source/numeric/SVInt.cpp

```diff
@@ -576,39 +586,26 @@ SVInt SVInt::ashr(const SVInt& rhs) const {
         return lshr(rhs);
     if (rhs.hasUnknown())
         return createFillX(bitWidth, signFlag);
-    if (rhs >= bitWidth)
+
+    auto amount = rhs.unsignedAmount();
+    if (amount >= bitWidth)
         return SVInt(bitWidth, *this >= 0 ? 0 : UINT64_MAX, signFlag);
 
-    return ashr(rhs.as<bitwidth_t>().value());
+    return ashr(amount);
 }
 
 SVInt SVInt::ashr(bitwidth_t amount) const {
-    if (!signFlag)
-        return lshr(amount);
     if (amount == 0)
         return *this;
     if (amount >= bitWidth)
         return SVInt(bitWidth, *this >= 0 ? 0 : UINT64_MAX, signFlag);
+    if (!signFlag || !(*this)[int32_t(bitWidth) - 1]) // unsigned or positive
```

Comment:
You can use isNegative() instead of checking the last bit manually.

Reply:
The purpose here is to detect isPositive(), or msb==0. isNegative()is msb==0. !isNegative() is msb== 0 or x or z.
So !isNegative() is not precisely isPositive(), when taking unknowns into consideration.

Reply:
Ah, I see. That's pretty subtle, can you expand the comment here to indicate what's happening?

Reply:
I add comments
```c++
    // !(*this)[int32_t(bitWidth) - 1]) checks msb==0
    // Not precisely !isNegative() which is msb!=1 or msb==[0xz]
    if (!signFlag || !(*this)[int32_t(bitWidth) - 1]) // unsigned or nonnegative
```